### PR TITLE
Drag&Drop Bugfixes and small refactoring

### PR DESCRIPTION
**Pull request by eyalhakim**
Fix bug onRemove emitted twice in OnDragStarted

Fix bug onAdding not called when dropping dragged tag

Reuse onAddingRequested inside onTagDropped and implement index support for onAddingRequested

### CODE REVIEW

File: modules/components/tag-input/tag-input.ts [hunk 1]
```diff
@@ -509,11 +509,8 @@ export class TagInputComponent extends TagInputAccessor implements OnInit, After
      * @name appendTag
      * @param tag {TagModel}
      */
-    public appendTag = (tag: TagModel): void => {
-        this.items = [
-            ...this.items,
-            this.modelAsStrings ? tag[this.identifyBy] : tag
-        ];
+    public appendTag = (tag: TagModel, index = this.items.length): void => {
+        this.items.splice(index, 0, tag);
```
Comment (Gbuomprisco) [inline modules/components/tag-input/tag-input.ts]: sorry, but doing so this.items gets mutated, I'd like something immutable instead

**Comment by Gbuomprisco** [comment]
Hi @eyalhakim,

thanks for the PR. but check out the tests as this broke a couple things

**Comment by eyalhakim** [comment]
Hi @Gbuomprisco i refactored as you requested and fixed the failing test, can you merge?

File: modules/components/tag-input/tag-input.ts [hunk 2]
```diff
@@ -509,11 +509,20 @@ export class TagInputComponent extends TagInputAccessor implements OnInit, After
...
+
+        for (let i = 0; i <= this.items.length; i++) {
+            if (i < index) {
+                items.push(this.items[i]);
+            } else if (i === index) {
+                items.push(this.modelAsStrings ? tag[this.identifyBy] : tag);
+            } else {
+                items.push(this.items[i - 1]);
+            }
+        }
```
Comment (Gbuomprisco) [inline modules/components/tag-input/tag-input.ts]: What is the reason of this change exactly? I cant understand from the code

Comment (eyalhakim) [inline modules/components/tag-input/tag-input.ts]: The original version would only insert the new tag at the end of the array, whereas in the case of drag&drop, the tag can be placed anywhere in the array.

Since this function is now used for drag&drop as well, it has to support an index.

Comment (eyalhakim) [inline modules/components/tag-input/tag-input.ts]: @Gbuomprisco i responded to you inside the file

Comment (Gbuomprisco) [inline modules/components/tag-input/tag-input.ts]: can you not reuse 
```
if (index === undefined) {		
             this.appendTag(tag);		
        } else {		
            const items = this.items;		
            this.items = [...items.slice(0, index), tag, ...items.slice(index, items.length)];		
        }
```

or does it do something wrong? I'm asking as this the latter seems more concise and easy to understand than a for loop

Comment (eyalhakim) [inline modules/components/tag-input/tag-input.ts]: i actually didn't need the if, it's from an older version.

Take a look now.

**Comment by eyalhakim** [comment]
@Gbuomprisco take a look now

File: modules/components/tag-input/tag-input.ts [hunk 3]
```diff
@@ -929,7 +917,7 @@ export class TagInputComponent extends TagInputAccessor implements OnInit, After
             .of(this.getItemDisplay(item))
             .map(display => this.setInputValue(display))
             .filter(display => this.inputForm.form.valid && !!display)
-            .map((display: string) => this.createTag(fromAutocomplete ? item : display))
+            .map((display: string) => this.createTag(item))
```
Comment (Gbuomprisco) [inline modules/components/tag-input/tag-input.ts]: as discussed - the items needs to undergo processing before it can be appended, which is why it needs to be display, rather than item, unless it comes from the autocomplete

Comment (Gbuomprisco) [inline modules/components/tag-input/tag-input.ts]: I understand there is a bug there, but we cannot introduce a further one to fix this so I suppose we need to handle it elsewhere. I hope to have time to take a look soon

Comment (eyalhakim) [inline modules/components/tag-input/tag-input.ts]: Oops, this one came in by mistake, i'm using it locally for now

**Comment by eyalhakim** [comment]
@Gbuomprisco please take a look again

**Comment by eyalhakim** [comment]
@Gbuomprisco Hi there, i reverted the change and yet the build still failed, don't see which test though.

Can you check?

File: modules/components/tag-input/tag-input.ts [hunk 4]
```diff
@@ -509,11 +509,9 @@ export class TagInputComponent extends TagInputAccessor implements OnInit, After
      * @name appendTag
      * @param tag {TagModel}
      */
-    public appendTag = (tag: TagModel): void => {
-        this.items = [
-            ...this.items,
-            this.modelAsStrings ? tag[this.identifyBy] : tag
```
Comment (Gbuomprisco) [inline modules/components/tag-input/tag-input.ts]: @eyalhakim here is the issue - if modelAsStrings is true, you need to append a string rather than an object

File: modules/components/tag-input/tag-input.ts [hunk 5]
```diff
@@ -509,11 +509,9 @@ export class TagInputComponent extends TagInputAccessor implements OnInit, After
...
-        ];
+    public appendTag = (tag: TagModel, index = this.items.length): void => {
+        const items = this.items;
```
Comment (Gbuomprisco) [inline modules/components/tag-input/tag-input.ts]: so maybe
```
const model = this.modelAsStrings ? tag[this.identifyBy] : tag;
this.items = [...items.slice(0, index), model, ...items.slice(index, items.length)]
```

Comment (eyalhakim) [inline modules/components/tag-input/tag-input.ts]: right! fixed

**Comment by eyalhakim** [comment]
@Gbuomprisco mergeable!
